### PR TITLE
feat(auth): Adding multi-tenancy and IdP management APIs

### LINKED ISSUE #353: FR: Multi Tenancy Support like in Node SDK
### [REQUIRED] Step 2: Describe your environment

  * Operating System version: MacOS Mojave 10.14.6
  * Firebase SDK version: 3.0
  * Library version: _____
  * Firebase Product: AUTH (auth, database, storage, etc)

### [REQUIRED] Step 3: Describe the problem

@hiranya911 
Would be terrific if we could support multi tenancy as it is in the Node sdk.
Is something like this in the pipeline?
I do think many users will love or benefit in having a multi-tenant support also for python.

### PR DESCRIPTION
**Pull request by hiranya911**
RELEASE NOTE: Added a new `tenant_mgt` module that provides APIs for creating, updating, retrieving, and deleting authentication tenants.
RELEASE NOTE: Added a new `tenant_mgt.auth_for_tenant()` function that returns a client object scoped to a specific tenant for performing tenant-aware user management operations.
RELEASE NOTE: Added the `auth.OIDCProviderConfig` type and related functions to create, retrieve, update and delete OIDC auth provider configurations.
RELEASE NOTE: Added the `auth.SAMLProviderConfig` type and related functions to create, retrieve, update and delete SAML auth provider configurations.

Resolves #353

### CODE REVIEW

**Comment by hiranya911** [comment]
@egilmorez please review the release notes, and the docstring chages in following files:

* auth.py
* _auth_client.py
* _auth_providers.py
* _user_mgt.py
* tenant_mgt.py

File: firebase_admin/_auth_client.py [hunk 1]
```diff
@@ -0,0 +1,625 @@
...
+            ValueError: If the user ID is None, empty or malformed.
+            FirebaseError: If an error occurs while deleting the user account.
+        """
+        self._user_manager.delete_user(uid)
+
+    def import_users(self, users, hash_alg=None):
+        """Imports the specified list of users into Firebase Auth.
+
+        At most 1000 users can be imported at a time. This operation is optimized for bulk imports
+        and will ignore checks on identifier uniqueness which could result in duplications. The
```
Comment (egilmorez) [inline firebase_admin/_auth_client.py:313]: Should this be "ignores checks on identifier uniqueness, which could result in duplications." or

"checks on identifier uniqueness that could result in duplications."

If there are some check that don't result in dupes, it should be the second one I think.  If all such checks result in dupes, the first.

File: firebase_admin/_auth_client.py [hunk 2]
```diff
@@ -0,0 +1,625 @@
...
+
+        Args:
+            provider_id: Provider ID string. Must have the prefix ``saml.``.
+            idp_entity_id: The SAML IdP entity identifier (optional).
+            sso_url: The SAML IdP SSO URL. Must be a valid URL (optional).
+            x509_certificates: The list of SAML IdP X.509 certificates issued by CA for this
+                provider  (optional).
+            rp_entity_id: The SAML relying party entity ID (optional).
+            callback_url: Callback URL string  (optional).
+            display_name: The user-friendly display name to the current configuration (optional).
```
Comment (egilmorez) [inline firebase_admin/_auth_client.py:567]: Unless this is technically required, "name of" sounds more natural.

File: firebase_admin/_auth_providers.py [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+    """Validates that the given value is a non-empty string."""
+    if not isinstance(value, str):
+        raise ValueError('Invalid type for {0}: {1}.'.format(label, value))
+    if not value:
+        raise ValueError('{0} must not be empty.'.format(label))
+    return value
+
+
+def _validate_url(url, label):
+    """Validates that the given value is a wellformed URL string."""
```
Comment (egilmorez) [inline firebase_admin/_auth_providers.py:371]: Should this just be "well-formed?"

File: firebase_admin/auth.py [hunk 4]
```diff
@@ -110,23 +123,23 @@
 UserRecord = _user_mgt.UserRecord
 
 
-def _get_auth_service(app):
-    """Returns an _AuthService instance for an App.
+def _get_client(app):
+    """Returns a Client instance for an App.
```
Comment (egilmorez) [inline firebase_admin/auth.py:127]: This seems like the sort of case where a literal isn't necessary, or maybe could be used just once.  Further references to "client instance" would work, to my ear anyway.  But it is also fine to repeat ``Client`` instance.

File: firebase_admin/auth.py [hunk 5]
```diff
@@ -527,54 +533,271 @@ def generate_sign_in_with_email_link(email, action_code_settings, app=None):
...
+
+def update_oidc_provider_config(
+        provider_id, client_id=None, issuer=None, display_name=None, enabled=None, app=None):
+    """Updates an existing OIDC provider config with the given parameters.
+
+    Args:
+        provider_id: Provider ID string. Must have the prefix ``oidc.``.
+        client_id: Client ID of the new config (optional).
+        issuer: Issuer of the new config (optional). Must be a valid URL.
+        display_name: The user-friendly display name to the current configuration (optional).
```
Comment (egilmorez) [inline firebase_admin/auth.py:605]: Suggest "name of"

File: firebase_admin/auth.py [hunk 6]
```diff
@@ -527,54 +533,271 @@ def generate_sign_in_with_email_link(email, action_code_settings, app=None):
...
+
+    Args:
+        provider_id: Provider ID string. Must have the prefix ``saml.``.
+        idp_entity_id: The SAML IdP entity identifier (optional).
+        sso_url: The SAML IdP SSO URL. Must be a valid URL (optional).
+        x509_certificates: The list of SAML IdP X.509 certificates issued by CA for this
+            provider  (optional).
+        rp_entity_id: The SAML relying party entity ID (optional).
+        callback_url: Callback URL string  (optional).
+        display_name: The user-friendly display name to the current configuration (optional).
```
Comment (egilmorez) [inline firebase_admin/auth.py:742]: Suggest "name of"

File: firebase_admin/tenant_mgt.py [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+        app=None):
+    """Updates an existing tenant with the given options.
+
+    Args:
+        tenant_id: ID of the tenant to update.
+        display_name: Updated display name string for the tenant (optional).
+        allow_password_sign_up: A boolean indicating whether to enable or disable the email sign-in
+            provider.
+        enable_email_link_sign_in: A boolean indicating whether to enable or disable email link
+            sign-in. Disabling this makes the password required for email sign-in.
```
Comment (egilmorez) [inline firebase_admin/tenant_mgt.py:130]: Is it really disabling it?  That seems mildly counterintuitive.

File: firebase_admin/_auth_client.py [hunk 8]
```diff
@@ -0,0 +1,625 @@
...
+
+class Client:
+    """Firebase Authentication client scoped to a specific tenant."""
+
+    def __init__(self, app, tenant_id=None):
+        if not app.project_id:
+            raise ValueError("""Project ID is required to access the auth service.
+            1. Use a service account credential, or
+            2. set the project ID explicitly via Firebase App options, or
+            3. set the project ID via the GOOGLE_CLOUD_PROJECT environment variable.""")
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply firebase_admin/_auth_client.py:36]: This isn't part of the documentation. It's a string constant in the code.

File: firebase_admin/_auth_client.py [hunk 9]
```diff
@@ -0,0 +1,625 @@
...
+            uid: A user ID string.
+
+        Raises:
+            ValueError: If the user ID is None, empty or malformed.
+            FirebaseError: If an error occurs while revoking the refresh token.
+        """
+        self._user_manager.update_user(uid, valid_since=int(time.time()))
+
+    def get_user(self, uid):
+        """Gets the user data corresponding to the specified user ID.
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply firebase_admin/_auth_client.py:135]: Seems alright to me. Let me know if I'm overlooking something obvious.

File: firebase_admin/_auth_client.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (egilmorez) [reply firebase_admin/_auth_client.py:36]: The step procedure isn't documentation?  I think I may not be putting comments on the right lines . . . but I figured GOOGLE_CLOUD_PROJECT was a literal.

File: firebase_admin/_auth_client.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (egilmorez) [reply firebase_admin/_auth_client.py:135]: Hm, maybe I missed "ID." because of a line break :)

File: firebase_admin/tenant_mgt.py (hunk 7 above)
Comment (hiranya911) [reply firebase_admin/tenant_mgt.py:130]: Setting it to False disables the feature

File: firebase_admin/auth.py (hunk 6 above)
Comment (hiranya911) [reply firebase_admin/auth.py]: Done

File: firebase_admin/auth.py (hunk 5 above)
Comment (hiranya911) [reply firebase_admin/auth.py]: Done

File: firebase_admin/auth.py (hunk 4 above)
Comment (hiranya911) [reply firebase_admin/auth.py]: Done. This is also an internal API and therefore is not included in the published docs.

File: firebase_admin/_auth_providers.py (hunk 3 above)
Comment (hiranya911) [reply firebase_admin/_auth_providers.py]: Done

File: firebase_admin/_auth_client.py (hunk 2 above)
Comment (hiranya911) [reply firebase_admin/_auth_client.py]: Done

File: firebase_admin/_auth_client.py (hunk 1 above)
Comment (hiranya911) [reply firebase_admin/_auth_client.py]: It's the first. Changed as suggested.

File: firebase_admin/_auth_client.py [hunk 10]
```diff
@@ -0,0 +1,625 @@
...
+        """
+        response = self._user_manager.get_user(phone_number=phone_number)
+        return _user_mgt.UserRecord(response)
+
+    def list_users(self, page_token=None, max_results=_user_mgt.MAX_LIST_USERS_RESULTS):
+        """Retrieves a page of user accounts from a Firebase project.
+
+        The ``page_token`` argument governs the starting point of the page. The ``max_results``
+        argument governs the maximum number of user accounts that may be included in the returned
+        page. This function never returns None. If there are no user accounts in the Firebase
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply firebase_admin/_auth_client.py]: Done

File: firebase_admin/_auth_client.py [hunk 11]
```diff
@@ -0,0 +1,625 @@
...
+        """Gets the user data corresponding to the specified phone number.
+
+        Args:
+            phone_number: A phone number string.
+
+        Returns:
+            UserRecord: A UserRecord instance.
+
+        Raises:
+            ValueError: If the phone number is None, empty or malformed.
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply firebase_admin/_auth_client.py]: None is a Python literal. Added code formatting

File: firebase_admin/_auth_client.py [hunk 12]
```diff
@@ -0,0 +1,625 @@
...
+        return _user_mgt.UserRecord(response)
+
+    def get_user_by_email(self, email):
+        """Gets the user data corresponding to the specified user email.
+
+        Args:
+            email: A user email address string.
+
+        Returns:
+            UserRecord: A UserRecord instance.
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply firebase_admin/_auth_client.py]: Agreed. Changing as suggested.

File: firebase_admin/_auth_client.py [hunk 13]
```diff
@@ -0,0 +1,625 @@
...
+        self._user_manager.update_user(uid, valid_since=int(time.time()))
+
+    def get_user(self, uid):
+        """Gets the user data corresponding to the specified user ID.
+
+        Args:
+            uid: A user ID string.
+
+        Returns:
+            UserRecord: A UserRecord instance.
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply firebase_admin/_auth_client.py]: Don't

File: firebase_admin/_auth_client.py [hunk 14]
```diff
@@ -0,0 +1,625 @@
...
+                    'Invalid tenant ID: {0}'.format(token_tenant_id))
+
+        if check_revoked:
+            self._check_jwt_revoked(verified_claims, _token_gen.RevokedIdTokenError, 'ID token')
+        return verified_claims
+
+    def revoke_refresh_tokens(self, uid):
+        """Revokes all refresh tokens for an existing user.
+
+        revoke_refresh_tokens updates the user's tokens_valid_after_timestamp to the current UTC
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply firebase_admin/_auth_client.py]: Done

File: firebase_admin/_auth_client.py [hunk 15]
```diff
@@ -0,0 +1,625 @@
...
+            ValueError: If input parameters are invalid.
+            TokenSignError: If an error occurs while signing the token using the remote IAM service.
+        """
+        return self._token_generator.create_custom_token(
+            uid, developer_claims, tenant_id=self.tenant_id)
+
+    def verify_id_token(self, id_token, check_revoked=False):
+        """Verifies the signature and data for the provided JWT.
+
+        Accepts a signed token string, verifies that it is current, and issued
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply firebase_admin/_auth_client.py]: Done

File: firebase_admin/_auth_client.py [hunk 16]
```diff
@@ -0,0 +1,625 @@
...
+from firebase_admin import _user_import
+from firebase_admin import _user_mgt
+
+
+class Client:
+    """Firebase Authentication client scoped to a specific tenant."""
+
+    def __init__(self, app, tenant_id=None):
+        if not app.project_id:
+            raise ValueError("""Project ID is required to access the auth service.
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply firebase_admin/_auth_client.py]: Done

**Comment by hiranya911** [comment]
@egilmorez made the suggested changes, and also did some additional changes to apply backticks where necessary. I think I've got nearly all of them.

File: firebase_admin/_auth_client.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (hiranya911) [reply firebase_admin/_auth_client.py:36]: Right. This whole block L33-L36 is an error message raised by the code. Not part of the documentation.

**Pull request merged**